File: Binary-Search/153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: list[int]) -> int:
        ans = float('inf')
        left, right = 0, len(nums) - 1

        while left <= right:
            mid = left + (right - left) // 2

            # If the array is already sorted
            if nums[left] <= nums[right]:
                ans = min(ans, nums[left])
                break

            # If the left part is sorted
            if nums[left] <= nums[mid]:
                ans = min(ans, nums[left])
                left = mid + 1
            # If the right part is sorted
            else:
                ans = min(ans, nums[mid])
                right = mid - 1

        return ans

File: Binary-Search/153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int ans = INT_MAX;
        int  left = 0, right = nums.size() - 1;

        while(left <= right){
            int mid = left + (right - left) / 2;

            if(nums[left] <= nums[right]){
                ans = min(ans, nums[left]);
            }

            if(nums[left] <= nums[mid]){
                ans = min(ans, nums[left]);
                left = mid + 1;
            }
            else{
                ans = min(ans, nums[mid]);
                right = mid - 1;
            }
        }

        return ans;
    }
};